cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java

```diff
@@ -1135,8 +1174,47 @@ private String getHiveDatabase(@Nullable String namespace) {
     if (namespace == null) {
       return null;
     }
-    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
-    return namespace.equals(Id.Namespace.DEFAULT.getId()) ? namespace : String.format("%s_%s", tablePrefix, namespace);
+    try {
+      if (namespace.equals(Id.Namespace.DEFAULT.getId())) {
+        return namespace;
+      } else {
+        String hiveDatabase = getCustomHiveDatabaseIfExists(namespace);
+        if (!Strings.isNullOrEmpty(hiveDatabase)) {
+          return hiveDatabase;
+        } else {
+          String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
+          return String.format("%s_%s", tablePrefix, namespace);
+        }
+      }
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Return the custom database for a given namespace if one exists else return null
+   *
+   * @param namespace the namespace for which we need to look up custom mapping
+   * @return the custom hive database name if one exists else null
+   * @throws ExploreException if failed to get NamespaceMeta for the given namespace
+   */
+  @Nullable
+  private String getCustomHiveDatabaseIfExists(@Nullable String namespace) throws ExploreException {
```

Comment:
I agree that its a little weird, since its being used both to find the db name and also to determine whether the db should be created/deleted. One possibility to clean it up is to have a getHiveDatabase method that returns an object containing the name of the database and a boolean indicating whether it needs to be created/deleted.